File: all/206_reverse_linked_list/sol_recursive.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseListRecursively(
    struct ListNode *previous_node,
    struct ListNode *current_node
) {
    struct ListNode *next_node = NULL;

    /*
     * Recursively change each current node to point to its previous node.
     * Termination condition: when current node is NULL, it means we have
     * traversed the whole linked list, and now the previous node will be the
     * new head.
     */
    if (NULL == current_node) {
        return previous_node;
    }

    /* Save the next node pointer. */
    next_node = (current_node->next);

    /* Change current node's next pointer to point to the previous node. */
    current_node->next = previous_node;

    return reverseListRecursively(current_node, next_node);
}

struct ListNode* reverseList(struct ListNode* head) {
    return reverseListRecursively(NULL, head);
}

File: all/206_reverse_linked_list/sol_recursive.py
from typing import Optional


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseListRecursively(
        self,
        previou_node: Optional[ListNode],
        current_node: Optional[ListNode]
    ) -> Optional[ListNode]:
        # Change each current node's next pointer to point to the previous node.
        # Termination condition: when current node is None, it means we have
        # traversed the whole linked list, and now previous node will be the
        # new head of the reversed linked list.
        if not current_node:
            return previou_node

        # Save the next node pointer.
        next_node = current_node.next

        # Change current node's next pointer to point to the previou node.
        current_node.next = previou_node

        return self.reverseListRecursively(current_node, next_node)

    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        return self.reverseListRecursively(None, head)
